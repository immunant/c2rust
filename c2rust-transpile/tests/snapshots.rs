use std::env::current_dir;
use std::fs;
use std::path::Path;
use std::process::Command;

use c2rust_transpile::{ReplaceMode, TranspilerConfig};

fn config() -> TranspilerConfig {
    TranspilerConfig {
        dump_untyped_context: false,
        dump_typed_context: false,
        pretty_typed_context: false,
        dump_function_cfgs: false,
        json_function_cfgs: false,
        dump_cfg_liveness: false,
        dump_structures: false,
        verbose: false,
        debug_ast_exporter: false,
        incremental_relooper: true,
        fail_on_multiple: false,
        filter: None,
        debug_relooper_labels: false,
        prefix_function_names: None,
        translate_asm: true,
        use_c_loop_info: true,
        use_c_multiple_info: true,
        simplify_structures: true,
        panic_on_translator_failure: false,
        emit_modules: false,
        fail_on_error: true,
        replace_unsupported_decls: ReplaceMode::Extern,
        translate_valist: true,
        overwrite_existing: true,
        reduce_type_annotations: false,
        reorganize_definitions: false,
        enabled_warnings: Default::default(),
        emit_no_std: false,
        output_dir: None,
        translate_const_macros: false,
        translate_fn_macros: false,
        disable_refactoring: false,
        preserve_unused_functions: false,
        log_level: log::LevelFilter::Warn,
        emit_build_files: false,
        binaries: Vec::new(),
    }
}

fn transpile(platform: Option<&str>, c_path: &Path) {
    let status = Command::new("clang")
        .args(&["-c", "-o", "/dev/null"])
        .arg(c_path)
        .status();
    assert!(status.unwrap().success());

    let (_temp_dir, temp_path) =
        c2rust_transpile::create_temp_compile_commands(&[c_path.to_owned()]);
    c2rust_transpile::transpile(config(), &temp_path, &[]);
    let cwd = current_dir().unwrap();
    let c_path = c_path.strip_prefix(&cwd).unwrap();
    let rs_path = c_path.with_extension("rs");
    let rs = fs::read_to_string(&rs_path).unwrap();
    let debug_expr = format!("cat {}", rs_path.display());

    let name = platform
        .map(|platform| ["transpile", platform].join("-"))
        .unwrap_or("transpile".into());

    insta::assert_snapshot!(name, &rs, &debug_expr);

    let status = Command::new("rustc")
        .args(&["--crate-type", "lib", "--edition", "2021", "-o", "-"])
        .arg(&rs_path)
        .status();
    assert!(status.unwrap().success());
}

#[test]
fn transpile_all() {
    insta::glob!("snapshots/*.c", |x| transpile(None, x));

    // Some things transpile differently on Linux vs. macOS,
    // as they use `unsigned long` and `unsigned long long` differently for builtins.
    // This makes snapshot tests trickier, as the output will be OS-dependent.
    // We handle this by adding OS name to the snapshot result filename.
    #[allow(unused)]
    let platform = "unknown";

    #[cfg(target_os = "linux")]
    let platform = "linux";
    #[cfg(target_os = "macos")]
    let platform = "macos";

    insta::with_settings!({snapshot_suffix => platform}, {
        insta::glob!("snapshots/platform-specific/*.c", |x| transpile(Some(platform), x));
    });
}
