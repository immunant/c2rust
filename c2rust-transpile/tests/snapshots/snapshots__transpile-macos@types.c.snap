---
source: c2rust-transpile/tests/snapshots.rs
expression: cat tests/snapshots/os-specific/types.macos.rs
input_file: c2rust-transpile/tests/snapshots/os-specific/types.c
---
#![allow(
    dead_code,
    non_camel_case_types,
    non_snake_case,
    non_upper_case_globals,
    unused_assignments,
    unused_mut
)]
pub type __darwin_ptrdiff_t = isize;
pub type __darwin_size_t = usize;
pub type __darwin_ssize_t = isize;
pub type int8_t = i8;
pub type int16_t = i16;
pub type int32_t = i32;
pub type int64_t = i64;
pub type intptr_t = isize;
pub type uintptr_t = usize;
pub type size_t = __darwin_size_t;
pub type uint8_t = u8;
pub type uint16_t = u16;
pub type uint32_t = u32;
pub type uint64_t = u64;
pub type intmax_t = ::libc::intmax_t;
pub type uintmax_t = ::libc::uintmax_t;
pub type ptrdiff_t = __darwin_ptrdiff_t;
pub type ssize_t = __darwin_ssize_t;
#[no_mangle]
pub static mut intvar: ::core::ffi::c_int = 0 as ::core::ffi::c_int;
#[no_mangle]
pub static mut sizevar: size_t = 0 as size_t;
#[no_mangle]
pub static mut ssizevar: ssize_t = 0 as ssize_t;
#[no_mangle]
pub static mut intptrvar: intptr_t = 0 as intptr_t;
#[no_mangle]
pub static mut uintptrvar: uintptr_t = 0 as uintptr_t;
#[no_mangle]
pub static mut ptrdiffvar: ptrdiff_t = 0 as ptrdiff_t;
#[no_mangle]
pub static mut uint8var: uint8_t = 0 as uint8_t;
#[no_mangle]
pub static mut uint16var: uint16_t = 0 as uint16_t;
#[no_mangle]
pub static mut uint32var: uint32_t = 0 as uint32_t;
#[no_mangle]
pub static mut uint64var: uint64_t = 0 as uint64_t;
#[no_mangle]
pub static mut int8var: int8_t = 0 as int8_t;
#[no_mangle]
pub static mut int16var: int16_t = 0 as int16_t;
#[no_mangle]
pub static mut int32var: int32_t = 0 as int32_t;
#[no_mangle]
pub static mut int64var: int64_t = 0 as int64_t;
#[no_mangle]
pub static mut maxvar: intmax_t = 0 as intmax_t;
#[no_mangle]
pub static mut umaxvar: uintmax_t = 0 as uintmax_t;

